Update social media modules

akohlbre opened the pull request:
- Adds support for blocking to the events module (to allow for a module to wait on the result of a listener)
- Updates `socialMediaActivity` to use the event model
- Updates `socialMediaLinkSharing` to use the new interface to `socialMediaActivity`
- Adds documentation in `socialMediaActivity`

Remaining tasks for these modules are in #17 and #23.

akohlbre committed:
create blocking events in socialMediaActivity

akohlbre committed:
documentation updates for socialMediaActivity

akohlbre committed:
Merge branch 'main' into socialmedia-events

jonathanmayer commented on src/events.js:
```diff
@@ -119,15 +119,19 @@ class Event {
     notifyListeners(listenerArguments = []) {
+        let returnVal = undefined;
         this.listeners.forEach((options, listener) => {
             try {
-                if((this.notifyListenersCallback === null) || this.notifyListenersCallback(listener, listenerArguments, options))
-                    listener.apply(null, listenerArguments);
+                if((this.notifyListenersCallback === null) || this.notifyListenersCallback(listener, listenerArguments, options)) {
+                    const listenerReturnValue = listener.apply(null, listenerArguments);
+                    if (listenerReturnValue !== undefined) { returnVal = listenerReturnValue; }
+                }
             }
             catch(error) {
                 debugLog(`Error in listener notification: ${error}`);
             }
         });
+        return returnVal;
```
Rather than clobbering the return value for each listener, so the `notifyListeners` caller only receives the last return value, what about storing all the return values in an array and returning that to the `notifyListeners` caller?

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -65,82 +66,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
+        registerPlatformListener(platform, eventType, blocking ? "blocking" : "nonblocking");
     }
-    tweetContentInit();
+    if (platform == 'twitter') tweetContentInit();
+    if (platform == 'facebook') fbPostContentInit();
```
=== seems right here.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -65,82 +66,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
- * @param blocking - whether the listener should be blocking. Allows canceling the event.
+ * Filter generated events to only the listeners for the specific event type.
+ * @param {socialMediaActivityCallback} listener - The listener that may be notified.
+ * @param {Array} listenerArguments - The event about to be sent to the listener.
```
Is there a reason to use an array here instead of just the argument?

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -65,82 +66,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
-    fbPostContentInit();
+function notifyFilter(listener, listenerArguments, options) {
+    const reportedEvent = listenerArguments[0];
+    const ret = reportedEvent['platform'] == options['platform'] &&
+        options['eventTypes'].includes(reportedEvent['eventType']);
```
`===` and `""`?

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -65,82 +66,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
+ * Options when adding a social media activity event listener.
+ * @typedef {Object} socialMediaActivityOptions
+ * @property {string} platform - The platform to add a listener for ('facebook', 'twitter', or 'reddit')
+ * @property {Array<string>} eventTypes - The events on the platform to notify for. For facebook, can
```
I've been trying to standardize on `string[]` for string array types.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -65,82 +66,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
-    }
-}
+
+/**
+ * @type {events.event<socialMediaActivityCallback, socialMediaActivityOptions}
+ */
```
See how events are annotated in other modules. We might eventually change the format, but it's the least-bad I could come up with for now.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -65,82 +66,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
+/**
+ * @type {events.event<socialMediaActivityCallback, socialMediaActivityOptions}
+ */
+export const onSocialMediaActivity = events.createEvent({
+    addListenerCallback: addListener,
+    notifyListenersCallback: notifyFilter
+});
```
A big-picture design question: do we want one event for all the platforms, or one event per platform? The advantage of the former is we avoid some repetition. The advantage of the latter is library users might shoot themselves in the foot by mixing and matching platforms and activity types.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -30,15 +31,15 @@ export function enablePrivateWindows() {
  * Unregistering is only necessary when there's already a nonblocking handler registered
  * and we want to convert it to a blocking handler.
  * @param platform - which social media platform the event is for
  * @param eventType - which type of event we're registering
```
Perhaps we should use the term "activity type" rather than "event type", since that's consistent with the name of the module and events are a specific concept in WebExtensions.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -82,82 +84,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
+/**
+ * Options when adding a social media activity event listener.
+ * @typedef {Object} socialMediaActivityOptions
+ * @property {string} platform - The platform to add a listener for ('facebook', 'twitter', or 'reddit')
```
`""` and .

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -82,82 +84,61 @@ function registerPlatformListener(platform, eventType, blockingType, callback) {
+
+/**
+ * Options when adding a social media activity event listener.
+ * @typedef {Object} socialMediaActivityOptions
```
I've been trying to establish a convention of lower camel case for callback types and upper camel case for other types (including `@callback` functions that aren't really callbacks).

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -776,6 +732,20 @@ function checkFacebookPostAudience(requestDetails) {
+ * within the object. This function is necessary to handle the frequest small changes that
+ * Facebook makes to its APIs. For example, they sometimes change a field from being a
+ * string to being an array of strings that only ever holds a single string, or vice versa.
+ * The function checks recurses down the object structure, unwinding fields as it goes:
```
"checks recurses" = ? And "unwinding fields" = ?

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -776,6 +732,20 @@ function checkFacebookPostAudience(requestDetails) {
 }
 
 /**
+ * Given an object and the name of a target field, attempt to find a field with that name
```
What does "field" mean?

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -733,7 +685,11 @@ function verifyFacebookComment({requestDetails = null}) {
 }
+
 /**
+ * Determine the audience of a Facebook post or reshare.
```
Here and throughout, when parsing platform-specific values, please make sure to document what the source data is and what assumptions you're making in parsing.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -852,6 +822,11 @@ async function extractFacebookReshare({requestDetails = null, verified = null, e
+ * This information is not included in the web request, nor is it attached to the generated post,
+ * so the only way to get it is by watching the user click the reshare button, which
+ * the content script does.
+ * @return {Promise} - The source of the last reshare ("person" or "page").
```
What type of Promise?

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -1113,6 +1092,11 @@ function extractRedditCommentVote({requestDetails = null, eventTime = null}) {
+/**
+ * For a given subreddit name, return whether it is private or public.
+ * @param {string} subredditName - The name of the subreddit, without leading "r/".
+ * @return {Promise} - resolves to a string representing the subreddit's status, or "unknown".
```
What type of Promise?

akohlbre committed:
handful of small docs/readability fixes to sMA

akohlbre committed:
document return value in notifyListeners

akohlbre committed:
Merge branch 'socialmedia-events' of github.com:mozilla-ion/web-science into socialmedia-events

akohlbre committed:
Merge branch 'main' into socialmedia-events

akohlbre committed:
split SMA into three events, add docs

standardize return values
add support for facebook reshares with 'include original post' option
add support for old reddit

akohlbre committed:
add support to SMLS for multiple different listeners

akohlbre commented:
This PR now contains all the code changes we're currently planning. I'll be pushing more documentation updates shortly, namely:
* Longer explanatory comments about usage at the top of SMA and SMLS
* Using the `@namespace` convention for documenting events and their associated object
* Clearer documentation for which fields are relevant to which event types in SMA
* More specific documentation about how we interact with platforms
* More specific documentation for a few return types (e.g. anything currently just `@returns {Object}`)

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -777,6 +946,26 @@ function checkFacebookPostAudience(requestDetails) {
+ * @param - the value of the fieldName, if found, or null.
  * @private
  */
 function findFieldFacebook(object, fieldName, enterArray = true, recurseLevel = 5) {
```
Would give a more descriptive name.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -822,46 +1011,69 @@ function findFieldFacebook(object, fieldName, enterArray = true, recurseLevel =
  * @private
  */
-async function getReshareInfo() {
+function getReshareInfo() {
     return browser.tabs.sendMessage(facebookTabId, {"recentReshare": true}).then((response) => {
         return response;
     }, (e) => { console.log("ERROR", e); } );
 }
```
Would use the `messaging` module and a schema. Would also give a more descriptive error.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -822,46 +1011,69 @@ function findFieldFacebook(object, fieldName, enterArray = true, recurseLevel =
  */
-async function getReshareInfo() {
+function getReshareInfo() {
     return browser.tabs.sendMessage(facebookTabId, {"recentReshare": true}).then((response) => {
```
This logic seems busted if the user has multiple Facebook pages open. Might be better to just send a message to the background script when there's a reshare and then consume the reshare event when notifying listeners.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -931,10 +1143,19 @@ function verifyFacebookReshare({requestDetails = null }) {
+ * @param {string} postId - the unique ID of the post
+ * @returns {facebookPostContentsDetails}
  */
 export function getFacebookPostContents(postId) {
     return new Promise((resolve, reject) => {
```
Same comment as the similar functionality for Twitter. This seems like it could be handled in the background page.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -949,79 +1170,167 @@ export function getFacebookPostContents(postId) {
 
 /**
  * Reddit posts don't currently have validation needs.
+ * @param {Object} requestDetails - the raw request.
+ * @returns {Object} - an empty object.
  * @private
  */
 function verifyRedditPost({requestDetails = null}) {
     return {};
 }
```
Seems fine to omit this.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -949,79 +1170,167 @@ export function getFacebookPostContents(postId) {
     return {};
 }
 
+/**
+ * Return a redditActivityDetails object with all fields initialized to default values.
+ * @returns {redditActivityDetails} - default-value Reddit details.
+ */
+function createBaseRedditObject() {
+    return {
+        eventType: "", eventTime: 0, postText: "", postAttachments: [],
+        subredditName: "", postTitle: "", voteDirection: 0, votedCommentId: "",
+        commentAttachments: [], commentText: "", commentedPostId: "", votedPostId: ""
+    };
+}
+
```
See comments about similar functions for Twitter and Facebook.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -949,79 +1170,167 @@ export function getFacebookPostContents(postId) {
 
-    details.postTitle = requestDetails.requestBody.formData.title[0];
-    details.eventType = "post";
+    let richtext_json = null;
```
See comment above about underscore variable names.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -1094,26 +1431,31 @@ function verifyRedditCommentVote({requestDetails = null}) {
+ */
 export function checkSubredditStatus(subredditName) {
     if (subredditName == "") return "unknown";
     return new Promise((resolve, reject) => {
```
Here and throughout, you don't need to wrap `fetch` in a new `Promise`, because it already starts a `Promise` chain.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -1139,17 +1481,49 @@ export function checkSubredditStatus(subredditName) {
     });
 }
 
+function addIfExists(sourceObject, destinationObject, sourceFieldName,
```
Documentation!

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -11,40 +11,78 @@ import * as socialMediaActivity from "./socialMediaActivity.js";
 /**
  * A callback function for the social media share event.
  * @callback socialMediaShareCallback
- * @param {SocialMediaShareDetails} details - Additional information about the social media share event.
+ * @param {socialMediaShareDetails} details - Additional information about the social media share event.
+ */
+
+/**
+ * An object giving information about a share on social media.
+ * @typedef {Object} socialMediaShareDetails
+ * @property {string} platform - the social media platform on which the user shared a link ("facebook",
+ *  "twitter", or "reddit").
+ * @property {number} shareTime - timestamp of the user's share.
+ * @property {boolean} tracked - whether this object gives information about a tracked link share or
+ *  about the count of untracked link shares.
+ *
+ * @property {string} eventType - for a tracked share, the action taken by the
+ *  user (e.g., "tweet", "reshare", etc).
+ * @property {string} url - for a tracked share, the URL that the user shared.
+ * @property {string} audience - for a tracked share, whether the share is visible to all ("public")
+ *  or not ("restricted"). When the privacy status cannot be determined, is set to "unknown".
+ * @property {string} reshareSource - for a tracked reshare on Facebook, whether the reshared
+ *  post came from a user's profile ("person"), or a company/organization ("page").
+ *
+ * @property {number} untrackedCount - the number of untracked links shared by a recent share event.
  */
 
 /**
  * Options when adding a social media share event listener.
- * @typedef {Object} SocialMediaShareOptions
- * @property {string[]} [matchPattern=[]] - The webpages of interest for the measurement, specified with WebExtensions match patterns.
+ * @typedef {Object} socialMediaShareOptions
+ * @property {string[]} [destinationMatchPatterns=[]] - The webpages of interest for the
+ *  measurement, specified with WebExtensions match patterns.
+ * @property {boolean} facebook - whether to measure shares that happen on Facebook.
+ * @property {boolean} twitter - whether to measure shares that happen on Twitter.
+ * @property {boolean} reddit - whether to measure shares that happen on Reddit.
+ */
```
See comments on Slack.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -11,40 +11,78 @@ import * as socialMediaActivity from "./socialMediaActivity.js";
  */
+const debugLog = debugging.getDebuggingLog("socialMediaLinkSharing");
+
+/**
+ * Unlike posts on Facebook, which have individual privacy controls, or posts on Reddit, whose
+ * privacy is determined by the subreddit in which they are posted, the privacy of a Tweet is
+ * determined by a global setting for each account. We pull this setting once and store it.
+ * @type {string}
+ * @private
+ */
+let twitterPrivacySetting = "unknown";
```
This seems like it should go in `socialMediaActivity`.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -60,13 +98,16 @@ function removeListener(listener) {
-    addListenerCallback: addListener,
-    removeListenerCallback: removeListener
-});
+function createBaseShareObject() {
+    return {
+        platform: "", shareTime: 0, url: "", audience: "", eventType: "", reshareSource: "",
+        type: "", untrackedCount: 0
+    };
+}
```
See other comments about these base object functions.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -84,16 +125,76 @@ async function startMeasurement({
     reddit = false
 }) {
```
Should replace the start and stop measurement conventions, since we don't use those anymore.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -84,16 +125,76 @@ async function startMeasurement({
     if (facebook) {
-        socialMediaActivity.registerFacebookActivityTracker(facebookLinks, ["post", "reshare"]);
+        socialMediaActivity.onFacebookActivity.addListener(facebookLinks, {
+            eventTypes: ['post', 'reshare']});
     }
     if (reddit) {
-        socialMediaActivity.registerRedditActivityTracker(redditLinks, ["post"]);
+        socialMediaActivity.onRedditActivity.addListener(redditLinks, {
+            eventTypes: ['post']});
     }
     if (twitter) {
-        socialMediaActivity.registerTwitterActivityTracker(twitterLinks, ["tweet", "retweet", "favorite"]);
+        socialMediaActivity.onTwitterActivity.addListener(twitterLinks, {
+            eventTypes: ['tweet', 'retweet', 'favorite']});
     }
```
Seems like some logic is needed in here for adding multiple listeners.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -84,16 +125,76 @@ async function startMeasurement({
+    return false;
+}
+
+/**
+ * Given events that were set up to be "tracked"-type notifications but do not match the current
+ *  listener's match patterns, turn them into one or more "untracked"-type notifications instead.
+ * @param {socialMediaShareDetails[]} trackedEvents - a set of non-matching share events.
+ * @returns {socialMediaShareDetails[]} - the tracked events stripped and condensed into untracked events.
+ * @private
+ */
+function trackedToUntracked(trackedEvents) {
```
Would just combine this with the prior function.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -84,16 +125,76 @@ async function startMeasurement({
+    return newEvents;
 }
 
 /**
```
Need to add logic for teardown after removing listeners.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -104,62 +205,73 @@ function stopMeasurement() {
  */
 function isTwitterLink(url) {
-    const twitterLink = /twitter\.com\/[0-9|a-z|A-Z|_]*\/status\/([0-9]*)\//;
+    const twitterLink = /twitter\.com\/[0-9|a-z|A-Z|_]*\/status\/([0-9]*)/;
```
Would parse to a `URL`, check the hostname, then check the path.

jonathanmayer commented on src/socialMediaLinkSharing.js:
```diff
@@ -11,40 +11,78 @@ import * as socialMediaActivity from "./socialMediaActivity.js";
+ * @type {Events.Event<socialMediaShareCallback, socialMediaShareOptions>}
  */
-const debugLog = debugging.getDebuggingLog("socialMediaLinkSharing");
+export const onShare = events.createEvent({
```
I think the boundary between `socialMediaActivity` and `socialMediaLinkSharing` should be a bit different from what it is. Currently, `socialMediaActivity` exposes functions for looking up tweets, posts, and subreddits, and `socialMediaLinkSharing` has to call those functions. It would be preferable if we could keep those functions internal to `socialMediaActivity` and only expose them if study developers eventually need them. The information that `socialMediaLinkSharing` has to obtain itself is also information that many callers to `socialMediaActivity` would want anyway. A possible approach would be a listener option for `socialMediaActivity` that indicates shared tweets and posts should be resolved, plus adding subreddit status to reddit events.

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -28,17 +29,155 @@ permissions.check({
+ * @property {number} eventTime - timestamp of user action.
+ * @property {string} tweetText - for a tweet, the content of the tweet.
+ * @property {string[]} tweetAttachments - for a tweet, any links attached to the tweet.
+ * @property {string} sharedId - for a favorite or share, the Twitter ID of the favorited or retweeted tweet.
```
Why not include a post ID for all Twitter, Facebook, and Reddit events?

jonathanmayer commented on src/socialMediaActivity.js:
```diff
@@ -28,17 +29,155 @@ permissions.check({
  */
```
The `permissions.check` call above should go in the add listener callback functions, with the right match pattern and permissions (`webRequestBlocking`?) for the event type and listener options.
